File: dawn_cpp/integer.cpp
#include <algorithm>
#include <vector>

class SignatureSort {
public:
    template <typename T>
    static std::vector<T> sort(const std::vector<T>& arr) {
        if (arr.size() < 2) {
            return arr;
        }

        std::vector<T> left(arr.begin(), arr.begin() + arr.size() / 2);
        std::vector<T> right(arr.begin() + arr.size() / 2, arr.end());

        left = sort(left);
        right = sort(right);

        return merge(left, right);
    }

private:
    template <typename T>
    static std::vector<T> merge(const std::vector<T>& left, const std::vector<T>& right) {
        std::vector<T> result;
        size_t i = 0, j = 0;

        while (i < left.size() && j < right.size()) {
            if (left[i] <= right[j]) {
                result.push_back(left[i++]);
            } else {
                result.push_back(right[j++]);
            }
        }

        result.insert(result.end(), left.begin() + i, left.end());
        result.insert(result.end(), right.begin() + j, right.end());

        return result;
    }
};

File: dawn_python/integer.py
class SignatureSort:
    def sort(self, arr):
        if len(arr) < 2:
            return arr

        mid = len(arr) // 2
        left = self.sort(arr[:mid])
        right = self.sort(arr[mid:])

        return self.merge(left, right)

    def merge(self, left, right):
        result = []
        i, j = 0, 0

        while i < len(left) and j < len(right):
            if left[i] <= right[j]:
                result.append(left[i])
                i += 1
            else:
                result.append(right[j])
                j += 1

        result.extend(left[i:])
        result.extend(right[j:])

        return result